### ISSUE TITLE
False-positive diag::err_undeclared_use_of_module possibly due to long paths

### ISSUE THREAD

**Issue by froody** [llvm/llvm-project#147220]
I recently started using clang-19.1.7 with bazel and I'm seeing errors sometimes like:

```
external/grpc+/src/core/load_balancing/oob_backend_metric.cc:30:10: error: module grpc+//:oob_backend_metric does not depend on a module exporting 'google/protobuf/duration.upb.h'
   30 | #include "google/protobuf/duration.upb.h"
      |          ^
```

When I look at the clang invocation I see:

`-fmodule-map-file=bazel-out/k8-fastbuild/bin/external/protobuf+/src/google/protobuf/duration_proto.upb.cppmap`

and the contents of that file is below. I suspect that the overuse of `../` in the paths generated by bazel is creating very long paths inside clang resulting in truncation, but I'm not sure how to validate this. 

Can someone familiar with module maps shed any light on this?

`duration_proto.upb.cppmap`:
```
module "protobuf+//src/google/protobuf:duration_proto.upb" {
  export *
  textual header "../../../../../../../../bazel-out/k8-fastbuild/bin/external/protobuf+/src/google/protobuf/_virtual_imports/duration_proto/google/protobuf/duration.upb.h"
  use "protobuf+//src/google/protobuf:duration_proto.upb_minitable"
  use "protobuf+//upb:generated_code_support__only_for_generated_code_do_not_use__i_give_permission_to_break_me"
  use "crosstool"
}
extern module "protobuf+//src/google/protobuf:duration_proto.upb_minitable" "../../../../../../../../bazel-out/k8-fastbuild/bin/external/protobuf+/src/google/protobuf/duration_proto.upb_minitable.cppmap"
extern module "protobuf+//upb:generated_code_support__only_for_generated_code_do_not_use__i_give_permission_to_break_me" "../../../../../../../../bazel-out/k8-fastbuild/bin/external/protobuf+/upb/generated_code_support__only_for_generated_code_do_not_use__i_give_permission_to_break_me.cppmap"
extern module "crosstool" "../../../../../../../../bazel-out/k8-fastbuild/bin/external/toolchains_llvm++llvm+llvm_toolchain/module-x86_64-linux.modulemap"
```

**Comment by toprakmurat** [comment]
Before implementing a full solution, you can test if path length is indeed the issue by creating shorter symlinks:

Example:

```bash
# Create a shorter symlink to test
ln -s bazel-out/k8-fastbuild/bin/external/protobuf+/src/google/protobuf/ short_protobuf

# Temporarily modify your module map to use the shorter path
# Replace the long path with something like:
# textual header "short_protobuf/duration.upb.h"
```

Also, It would be helpful if you could also share what you have tried so far and what didn’t work.

**Comment by froody** [comment]
It turns out this error above was not path length but rather a -isystem vs -iquote due to a bazel change.

I did however verify a similar issue on macos was fixed by shortening the paths. I basically renamed the directory containing the .cppmap files to a shorter name and replaced all the references in the .cppmap files, and then I was able to successfully compile.

I'm happy to recompile clang and investigate myself, I just don't know where to start adding logging/etc.

**Comment by TomMurray** [comment]
I encountered what I believe is this same issue in an internal project using bazel and decided to dig into it.
Managed to cut it down to a reasonably minimal reproducer: [repro.zip](https://github.com/user-attachments/files/27003765/repro.zip)

Essentially as OP described: `extern module` declarations with relative paths are just concatenated together, and bazel produces a lot of long chains of `../../` due to reaching back up to the execroot or something. `extern module` paths are allowed to not be found because they are lazily evaluated, so clang intentionally ignores a failure to find the file. In this case the file path is not found because it grows too long (all the repeated `../../path/path`) and hits OS path limits (particularly on macOS where `PATH_MAX` is shorter I believe) meaning it fails to be parsed and causes the unexpected module not found error when evaluating.

I've put up a suggested fix for the unbounded path growth with relative module paths: #193691. This doesn't really fix the fact that unexpected errors here are not reported well but my understanding of modules isn't good enough to say when it's safe to report those errors due to the lazy evaluation thing mentioned above.

**Comment by jyknight** [comment]
In Google's internal use, we dealt with this issue since the very beginning by using source-root-relative modulemaps. That's enabled in bazel cc toolchain by enabling the feature "module_map_home_cwd", and in Clang by passing `-Xclang=-fmodule-map-file-home-is-cwd`. I've never been terribly happy with that solution, since it's using an internal flag and also changing the interpretation of _all_ modulemaps, but it's been the status quo for a decade.
